### ISSUE TITLE
synchronize of ChannelListController is very slow when a user has joined 20+ channels and each channel has 30+ members.

### ISSUE THREAD

**Issue by salqueng** [GetStream/stream-chat-swift#1823]
## What did you do?
Connect and synchronize the joined channel list of a user.
Tested on the playground of this repo.

The user has joined 20+ channels, and each channel has more than 30 members.

<details>
<summary>Playground Code</summary>
<p>

```swift
class IgnoreDataSavedDestination: BaseLogDestination {
  override func write(message: String) {
    // For hiding long "Data saved to DB" logs
    if message.contains("Data saved to DB") {
      return
    }
    print(message)
  }
}

LogConfig.level = .debug
LogConfig.subsystems = [.database]
LogConfig.destinationTypes = [IgnoreDataSavedDestination.self]

/* Configuration */
let apiKeyString = "[REDACTED]"
let userId = "[REDACTED]"
let token: Token = "[REDACTED]"

let client = StreamChat.ChatClient(
  config: .init(apiKeyString: apiKeyString)
)

client.connectUser(userInfo: .init(id: userId), token: token) { error in
  if let error = error {
    print(error)
    return
  }
  print(LogConfig.dateFormatter.string(from: Date()), "---Connected---")

  let controller = client.channelListController(
    query: .init(
      filter: .containMembers(userIds: [userId])
    )
  )

  controller.synchronize { error in
    if let error = error {
      print(error)
      return
    }
    print(LogConfig.dateFormatter.string(from: Date()), "---Synchronized---", controller.channels.count)
  }
}

```
</p>
</details>

## What did you expect to happen?
Synchronize finishes within 1 ~ 2 seconds so users can feel the app is quite fast.

## What happened instead?
Synchronize finishes after 7 ~ 9 seconds.

Here is the log - You can see the database operations take several seconds.

<details>
<summary>Log</summary>
<p>

```
2022-03-03 13:06:05.400 [DEBUG] [main] [DatabaseContainer.swift:320] [resetEphemeralValues()] > Ephemeral values reset.
2022-03-03 13:06:05.714 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-03-03 13:06:05.716 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-03-03 13:06:05.718 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-03-03 13:06:05.718 ---Connected---
2022-03-03 13:06:05.861 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-03-03 13:06:05.862 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-03-03 13:06:05.862 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-03-03 13:06:05.951 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-03-03 13:06:05.951 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:198] [write(_:completion:)] > Context has no changes. Skipping save.
2022-03-03 13:06:05.951 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-03-03 13:06:06.645 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-03-03 13:06:06.646 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-03-03 13:06:06.647 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-03-03 13:06:08.847 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-03-03 13:06:13.315 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-03-03 13:06:13.503 ---Synchronized--- 20
2022-03-03 13:06:13.503 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-03-03 13:06:13.835 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-03-03 13:06:13.850 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-03-03 13:06:13.858 [DEBUG] [NSManagedObjectContext 0x600003b381a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
```

</p>
</details>

Setting smaller pageSize, messagesLimit, and membersLimit speeds up synchronization, but this might hurt user experience.

## GetStream Environment
**GetStream Chat version: 4.10.1**
**GetStream Chat frameworks: StreamChat**
**iOS version: 15.2**
**Swift version: 5.5.2**
**Xcode version: 13.2.1**
**Device: iPhone 13 Simulator**

## Additional context
- REST API isn't slow (Android and JS SDK do not have above problems)

**Comment by b-onc** [comment]
Hello @salqueng ,

`Synchronize` completion shouldn't take even a second. Can you test it like this:

```swift
log.warning("Starting sync")
controller.synchronize { _ in
  log.warning("Sync finished")
}
```
and report the timestamps?

**Comment by salqueng** [comment]
Hmm. It actually takes several seconds for my cases. You can see the test code in "Playground Code" section of first comment.

<img width="861" alt="image" src="https://user-images.githubusercontent.com/5074074/156603862-e4599572-33ce-452a-801d-996ff37c9e68.png">

I can send you the credentials for the above code.

**Comment by b-onc** [comment]
Hello @salqueng ,

Thank you. Can you send me Stream app id (not API key, you can find your app id via dashboard), user id, and the channel query causing this?

**Comment by salqueng** [comment]
Thank you @b-onc,
Here is the information.

App Id: `"1126626"`
User Id: `"190176"`
Channel Query: (in swift) `.containMembers(userIds: ["190176"])`

**Comment by b-onc** [comment]
Hello @salqueng ,

Thanks for the information. I'm able to replicate, and even pinpoint, the performance bottleneck. We'll prioritize it and hope to achieve 50%+ improvements soon.

However, I can't guarantee a timeline for an improvement so for now I suggest you to decrease channel/message limits for your query to make it faster. Also, make sure your `baseURL` points to `singapore`

**Comment by tbarbugli** [comment]
@salqueng there is some initial work to improve performance around your use-case, if you have time we'd appreciate if you could give the `perf-coredata` branch a spin and let us know if you see an improvement

**Comment by salqueng** [comment]
<details>
<summary>Tested on 4.13.1: Takes 8.292s</summary>
<p>

```
2022-04-20 22:27:26.877 [DEBUG] [main] [DatabaseContainer.swift:320] [resetEphemeralValues()] > Ephemeral values reset.
2022-04-20 22:27:27.211 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-04-20 22:27:27.213 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:27:27.215 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:27:27.215 ---Connected---
2022-04-20 22:27:27.368 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-04-20 22:27:27.368 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:198] [write(_:completion:)] > Context has no changes. Skipping save.
2022-04-20 22:27:27.368 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:27:27.459 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-04-20 22:27:27.460 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:198] [write(_:completion:)] > Context has no changes. Skipping save.
2022-04-20 22:27:27.460 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:27:28.130 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-04-20 22:27:28.135 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:27:28.136 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:27:28.970 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-04-20 22:27:28.971 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:27:28.972 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:201] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:27:30.583 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:182] [write(_:completion:)] > Starting a database session.
2022-04-20 22:27:35.309 [DEBUG] [NSManagedObjectContext 0x6000032201a0] [DatabaseContainer.swift:195] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:27:35.507 ---Synchronized--- 20
```

</p>
</details>

<details>
<summary>Tested on #1937: Takes 3.502s</summary>
<p>

```
2022-04-20 22:20:04.610 [DEBUG] [main] [DatabaseContainer.swift:325] [resetEphemeralValues()] > Ephemeral values reset.
2022-04-20 22:20:05.087 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:186] [write(_:completion:)] > Starting a database session.
2022-04-20 22:20:05.093 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:200] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:20:05.096 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:206] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:20:05.096 ---Connected---
2022-04-20 22:20:05.147 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:186] [write(_:completion:)] > Starting a database session.
2022-04-20 22:20:05.147 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:203] [write(_:completion:)] > Context has no changes. Skipping save.
2022-04-20 22:20:05.147 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:206] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:20:05.270 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:186] [write(_:completion:)] > Starting a database session.
2022-04-20 22:20:05.270 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:203] [write(_:completion:)] > Context has no changes. Skipping save.
2022-04-20 22:20:05.270 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:206] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:20:05.445 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:186] [write(_:completion:)] > Starting a database session.
2022-04-20 22:20:05.461 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:200] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:20:05.464 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:206] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:20:05.464 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:186] [write(_:completion:)] > Starting a database session.
2022-04-20 22:20:05.464 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:200] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:20:05.465 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:206] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:20:06.247 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:186] [write(_:completion:)] > Starting a database session.
2022-04-20 22:20:06.251 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:200] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:20:06.254 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:206] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:20:08.542 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:186] [write(_:completion:)] > Starting a database session.
2022-04-20 22:20:08.592 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:200] [write(_:completion:)] > Context has changes. Saving.
2022-04-20 22:20:08.594 [DEBUG] [NSManagedObjectContext 0x600002db56c0] [DatabaseContainer.swift:206] [write(_:completion:)] > Database session succesfully saved.
2022-04-20 22:20:08.598 ---Synchronized--- 20
```

</p>
</details>

It's definitely improved! But I didn't check the channels are loaded correctly.

**Comment by tbarbugli** [comment]
@salqueng thank you for sharing this, would you mind running this a couple times in your shell and attach its output here?

```
time curl 'https://chat.stream-io-api.com/channels' --verbose
```

**Comment by salqueng** [comment]
Takes about 200ms

<details>
  <summary>Result</summary>
<p>

```
*   Trying 35.72.92.82:443...
* Connected to chat.stream-io-api.com (35.72.92.82) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=*.stream-io-api.com
*  start date: Nov 18 00:00:00 2021 GMT
*  expire date: Dec 17 23:59:59 2022 GMT
*  subjectAltName: host "chat.stream-io-api.com" matched cert's "*.stream-io-api.com"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x135812200)
> GET /channels HTTP/2
> Host: chat.stream-io-api.com
> user-agent: curl/7.79.1
> accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 2147483647)!
< HTTP/2 401
< www-authenticate: Bearer realm="QueryChannels failed with error: \"api_key not provided\""
< content-type: application/json
< content-length: 176
< date: Mon, 02 May 2022 10:09:37 GMT
< server: stream-edge
<
* Connection #0 to host chat.stream-io-api.com left intact
{"code":5,"message":"QueryChannels failed with error: \"api_key not provided\"","StatusCode":401,"duration":"","more_info":"https://getstream.io/chat/docs/api_errors_response"}curl 'https://chat.stream-io-api.com/channels' --verbose  0.02s user 0.01s system 11% cpu 0.247 total
```

</p>
</details>

**Comment by polqf** [comment]
Hi @salqueng,

We have been (and we still are) working on a series of improvements to reduce the times when saving big chunks of data to the database. We've recently merged a PR that is giving a 50% improvement in our tests. You can check the CHANGELOG here: https://github.com/GetStream/stream-chat-swift/blob/develop/CHANGELOG.md

We'll keep you posted. Thank you!

**Comment by evsaev** [comment]
Hi @salqueng 👋 

I'm happy to say that we've released [4.16.0](https://github.com/GetStream/stream-chat-swift/releases/tag/4.16.0) that contains the performance improvement we mentioned earlier :tada:

I'm gonna close the issue but feel free to reopen in case you face the performance drop in this area again.

Thank you for your patience!

Best,
Pavel
